### PR TITLE
Add pycvodes, pygslodeiv2, pyodeint

### PR DESCRIPTION
**Pull request by bjodah**
These are some packages of mine. They have a few users (nothing compared to e.g. `scipy`), are they within scope of `conda-forge`? Or do packages hosted on conda-forge need to have a wide user-base to qualify?

P.S. Windows is not yet supported but I will add support for that at a later stage.

### CODE REVIEW

**Comment by bjodah** [comment]
so the compilers used by default on conda-forge do not yet support C++14.

**Comment by moorepants** [comment]
That's right. I hit that on the simbody feedstock. Why is C++14 needed?

**Comment by bjodah** [comment]
It's not strictly needed, but instead of rewriting as C++11 (having to provide `make_unique<double[]>` etc.) I would actually want to move on to C++17 (`std::aligned_alloc`, `std::optional` etc.)...

I'm refactoring the underlying lib right now anyway so could try to make it C++11.

**Comment by bjodah** [comment]
@moorepants -- there you go :)

**Comment by moorepants** [comment]
Wow! You are awesome! Thanks.

+1 to merge :)

File: recipes/pycvodes/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+requirements:
+  build:
+    - blas 1.1 {{ variant }}  # [not win]
+    - openblas 0.2.20*  # [not win]
+    - sundials >=3.1.0
+    - python
+    - setuptools
+    - pip
+    - numpy
```
Comment (isuruf) [inline recipes/pycvodes/meta.yaml]: Is this using numpy C API?

File: recipes/pycvodes/build.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+export PYCVODES_LAPACK=openblas
+export PYCVODES_SUNDIALS_LIBS=sundials_cvodes,sundials_nvecserial,sundials_sunlinsollapackdense,sundials_sunlinsollapackband
+cat <<EOF>pycvodes/_config.py
+env = {
+    'LAPACK': "${PYCVODES_LAPACK}",
+    'SUNDIALS_LIBS': "${PYCVODES_SUNDIALS_LIBS}"
+}
+EOF
+export PYCVODES_STRICT=1
+export CPLUS_INCLUDE_PATH=${PREFIX}/include
```
Comment (isuruf) [inline recipes/pycvodes/build.sh]: Add `toolchain` to build requirements and you can remove this.

File: recipes/pycvodes/meta.yaml (hunk 1 above)
Comment (bjodah) [reply recipes/pycvodes/meta.yaml]: I believe so (I have Cython-function signatures such as: `cdef _reshape_roots(cnp.ndarray[cnp.float64_t, ndim=1] roots, int ny):`)

Comment (isuruf) [reply recipes/pycvodes/meta.yaml]: See https://conda-forge.org/docs/meta.html#building-against-numpy

File: recipes/pycvodes/meta.yaml [hunk 3]
```diff
@@ -27,13 +27,13 @@ requirements:
...
     - pip
-    - numpy
+    - numpy 1.11.*
   run:
     - blas 1.1 {{ variant }}  # [not win]
     - openblas 0.2.20|0.2.20.*  # [not win]
     - sundials >=3.1.0
     - python
-    - numpy
+    - numpy numpy>=1.11
```
Comment (isuruf) [inline recipes/pycvodes/meta.yaml]: typo

File: recipes/pygslodeiv2/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [win]
+  skip: True  # [osx and py27]
+  script: ${PYTHON} -m pip install --no-deps --ignore-installed .
```
Comment (isuruf) [inline recipes/pygslodeiv2/meta.yaml]: Use `python` instead of `${PYTHON}`

**Comment by bjodah** [comment]
Look like Circle CI stalled, would you mind trying to restart?

**Comment by moorepants** [comment]
+1 to merge this (second time)

Also, @bjodah do your packages build on windows or is there a lot to do to get that to happen. I see that sundials is available on windows via conda-forge.

File: recipes/pygslodeiv2/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  script: python -m pip install --no-deps --ignore-installed .
+
+requirements:
+  build:
+    - toolchain
+    - python
+    - setuptools
+    - pip
+    - numpy 1.11.*
+    - gsl
```
Comment (isuruf) [inline recipes/pygslodeiv2/meta.yaml]: Are you linking to openblas with gsl?

Comment (bjodah) [reply recipes/pygslodeiv2/meta.yaml]: Indirectly?:
https://github.com/conda-forge/gsl-feedstock/blob/master/recipe/meta.yaml#L36
https://github.com/bjodah/pygslodeiv2/blob/master/pygslodeiv2/_config.py#L2
https://github.com/bjodah/pygslodeiv2/blob/master/setup.py#L59
so I link against `-lgslcblas`, I could change that by setting `PYGSLODEIV2_BLAS=openblas`.
(sorry, I'm a bit fuzzy on exactly what linkers do when linking against a shared library which was linked against another one...)

Comment (isuruf) [reply recipes/pygslodeiv2/meta.yaml]: In conda-forge gslcblas is a link to openblas. So you need to pin openblas here. It's better to add `PYGSLODEIV2_BLAS=openblas` as well.

**Comment by bjodah** [comment]
@moorepants in short: no. I tried at some point but I'm unfamiliar with compiling things on windows and haven't found the time since to give it another try. That said, I can't think of any reason for why it shouldn't work.

**Comment by moorepants** [comment]
Ok, I may give it a shot for windows if I can make some time once this is merged.

**Comment by moorepants** [comment]
@isuruf Can this go in?

**Comment by moorepants** [comment]
ping, curious is this is good to merge? seems like @bjodah has addressed @isuruf's comments.

File: recipes/pycvodes/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+build:
+  number: 200
+  skip: True  # [win]
+  features:
+    - blas_{{ variant }}  # [not win]
+
+requirements:
+  build:
+    - toolchain
```
Comment (isuruf) [inline recipes/pycvodes/meta.yaml]: @bjodah, can you modify the recipe to use conda-build=3 stuff? Here's an experimental script https://gist.github.com/isuruf/a4c4c0a0c49114bc02cd8554b5d3b1f7

**Comment by isuruf** [comment]
Sorry, was a bit busy with exams last week. I think this can be merged after changing to conda-build=3 features.

File: recipes/pycvodes/meta.yaml (hunk 6 above)
Comment (bjodah) [reply recipes/pycvodes/meta.yaml]: Are there any examples of how `conda_build_config.yaml` should look like on conda-forge?
For example, does this look alright?:
```
# Pinning packages
python:
    - 2.7
    - 3.5
    - 3.6
numpy:
    - 1.11
openblas:
    - 0.2.20
```
(and is the `# Pinning packages` in the right place for use with your script?)

**Comment by moorepants** [comment]
Any reason this can't be merged as is and then upgrade to the new features after the recipe is in?

File: recipes/pycvodes/meta.yaml (hunk 6 above)
Comment (isuruf) [reply recipes/pycvodes/meta.yaml]: Use the config file from `conda-forge-pinning`

**Comment by bjodah** [comment]
Now it's pulling sundials 2.7.0 instead of 3.1.0 during build:
```
sundials:        2.7.0-blas_openblas_5        conda-forge [blas_openblas]
```
(since the re-render script dropped >=3.1.0)
1. I never understood why 2.7.0 was preferred to begin with (I figured something with dependency resolution)
2. Now I'm unsure where this restriction should go, should I readd the requirement to the `meta.yaml` file?

**Comment by isuruf** [comment]
@bjodah, please send a PR to fix the line https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml#L373-L374

**Comment by bjodah** [comment]
OK, so sundials cannot be pinned higher than 2.7 due to use of assimulo.
Should I:
- Build against sundials 2.7
or should I:
- specify sundials >= 3.1
in `meta.yaml`?

**Comment by jakirkham** [comment]
It may be possible to do both if it’s of interest.

**Comment by jakirkham** [comment]
As to the question in the OP, we have no rules on this. We rely on the community to package software that they think is useful.

**Comment by bjodah** [comment]
I linked against sundials 2.7 for now. Let me know if you want to see any more changes to any of the recipes.

**Comment by moorepants** [comment]
Looks like this merge failed on master.

**Comment by moorepants** [comment]
```
Repository created, please edit conda-forge.yml to configure the upload channels
and afterwards call 'conda smithy register-github'
Traceback (most recent call last):
  File ".travis_scripts/create_feedstocks.py", line 146, in <module>
    if repo_exists(gh, 'conda-forge', name + '-feedstock'):
  File ".travis_scripts/create_feedstocks.py", line 56, in repo_exists
    org = gh.get_organization(organization)
  File "/home/travis/miniconda/lib/python2.7/site-packages/github/MainClass.py", line 244, in get_organization
    "/orgs/" + login
  File "/home/travis/miniconda/lib/python2.7/site-packages/github/Requester.py", line 185, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, cnx))
  File "/home/travis/miniconda/lib/python2.7/site-packages/github/Requester.py", line 198, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.RateLimitExceededException: 403 {u'documentation_url': u'https://developer.github.com/v3/#rate-limiting', u'message': u'API rate limit exceeded for user ID 19853139.'}
+++python .travis_scripts/trigger_travis_build.py conda-forge/staged-recipes
{
  "@type": "error",
  "error_type": "request_limit_reached",
  "error_message": "request limit reached for resource",
  "repository": {
    "@type": "repository",
    "@href": "/repo/5842170",
    "@representation": "minimal",
    "id": 5842170,
    "name": "staged-recipes",
    "slug": "conda-forge/staged-recipes"
  },
  "max_requests": 10,
  "per_seconds": 3600
}
Traceback (most recent call last):
  File ".travis_scripts/trigger_travis_build.py", line 54, in <module>
    sys.exit(main(sys.argv))
  File ".travis_scripts/trigger_travis_build.py", line 47, in main
    rebuild_travis(args.slug)
  File ".travis_scripts/trigger_travis_build.py", line 38, in rebuild_travis
    response.raise_for_status()
  File "/home/travis/miniconda/lib/python2.7/site-packages/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 429 Client Error: Too Many Requests for url: https://api.travis-ci.org/repo/conda-forge%2Fstaged-recipes/requests
```

**Pull request merged**